### PR TITLE
PR: Add delay and minimum chars to on the fly completions

### LINKED ISSUE #10134: On the fly code completion interferes with normal typing
### What steps will reproduce the problem?

This may be intended behavior, however, it didn't happen with 3.X. 

With code completion and on the fly code completion turned on (note these are on different panes in preferences, which makes discovery hard), try to type the following function definition in a new editor window.

    def f(x):
        return x

The code completion pops up a window with xrange as the completion of the `return x`. If you are typing quickly (as I do) it will auto-complete x as xrange, yielding

    def f(x):
        return xrange

I think 3.x had a way to control whether ENTER auto completes or not, I don't think I see that in 4 (this issue reporter window is modal, so I can't check while I am typing this)

Anyway, this may be the intended/desired behavior, but I find it hard to imagine most will find it useful. One may counter that x is a bad variable name, but I only really do math in Python, so x is very common usage. 

## Versions

* Spyder version: 4.0.0b4 
* Python version: 3.7.3
* Qt version: 5.9.6
* PyQt5 version: 5.9.2
* Operating System: Windows 7

### Dependencies

```
pygments >=2.0    :  2.4.2 (OK)
qdarkstyle >=2.6.4:  2.7 (OK)
sphinx >=0.6.6    :  2.1.2 (OK)
pyls >=0.27.0     :  0.28.2 (OK)
nbconvert >=4.0   :  5.5.0 (OK)
pandas >=0.13.1   :  0.25.1 (OK)
numpy >=1.7       :  1.16.4 (OK)
sympy >=0.7.3     :  1.4 (OK)
cython >=0.21     :  0.29.13 (OK)
qtconsole >=4.5.2 :  4.5.4 (OK)
IPython >=4.0     :  7.8.0 (OK)
matplotlib >=2.0.0:  3.1.1 (OK)
pylint >=0.25     :  2.3.1 (OK)
```

### PR DESCRIPTION
**Pull request by goanpeca**
## Description of Changes

- Add configurable time delay to on the fly completion
- Add configurable characters after which on the fly completion is triggered
- Remove on the fly completion for strings and/or comments (normal `tab` completion will still work)
- Split configuration options and add a new tab 

* [x] Wrote at least one-line docstrings (for any new functions)

image

image

image

### Issue(s) Resolved

Fixes #10134

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @goanpeca

### CODE REVIEW

**Comment by goanpeca** [comment]
Screen Shot 2019-09-19 at 23 39 31
Screen Shot 2019-09-19 at 23 39 38
Screen Shot 2019-09-19 at 23 39 46

@ccordoba12

**Comment by goanpeca** [comment]
@CAM-Gerlach could you try this one?

**Comment by CAM-Gerlach** [comment]
@goanpeca I tested it out and the character, time delay, string and comment no-completion, and settings all worked very well, and the preferences looked great, and I didn't get any errors during limited testing. There may be a few corner cases (I don't have much time at the moment to test with my usual level of exhaustiveness) but overall its a huge improvement from the previous state of affairs.

There's one thing, though—since you already have the option implemented, can you allow a minimum delay of 0 ms rather than 200 ms, for users who prefer maximum responsiveness in their OTF autocompletion? As far as I'm aware there's no technical reason to restrict it (given it was 0 before, and its already configurable so it doesn't increase complexity any further) and some users may not want any delay or very little (I had it set to 100 ms in Rstudio) though 300 ms is more than fine as the default.

In any case, I'll certainly be making great use of OTF autocompletion thanks to this patch as well as all the previous work done on it. Aside from the handful of remaining errors I regularly run into, as a user I can say I'm very pleased with how things have turned out in Spyder 4 in this department, and you all should be very proud of your work. Thank you!

File: spyder/plugins/completion/languageserver/confpage.py [hunk 1]
```diff
@@ -692,17 +692,59 @@ class LanguageServerConfigPage(GeneralConfigPage):
     def setup_page(self):
         newcb = self.create_checkbox
 
-        # --- Introspection ---
-        # Basic features group
-        basic_features_group = QGroupBox(_("Basic features"))
-        completion_box = newcb(_("Enable code completion"), 'code_completion')
+        # --- Completion ---
+        # Completion group
+        completion_label = QLabel(_('Some text for this section!'))
```
Comment (ccordoba12) [inline spyder/plugins/completion/languageserver/confpage.py]: There's no need for a label in this section, so please remove this one.

File: spyder/plugins/completion/languageserver/confpage.py [hunk 2]
```diff
@@ -692,17 +692,59 @@ class LanguageServerConfigPage(GeneralConfigPage):
...
+        self.completions_after_characters = self.create_spinbox(
+            _("Show automatic completions after characters:"), None,
+            'automatic_completions_after_chars', min_=1, step=1,
+            tip=_("Default is 3"), section='editor')
+        self.completions_after_ms = self.create_spinbox(
+            _("Show automatic completions after keyboard idle (ms):"), None,
+            'automatic_completions_after_ms', min_=200, max_=5000, step=10,
+            tip=_("Default is 300"), section='editor')
         code_snippets_box = newcb(_("Enable code snippets"), 'code_snippets')
         enable_hover_hints_box = newcb(
```
Comment (ccordoba12) [inline spyder/plugins/completion/languageserver/confpage.py]: Please move this checkbox to the `Introspection` tab because it's not related to code completion.

File: spyder/plugins/completion/languageserver/confpage.py [hunk 3]
```diff
@@ -692,17 +692,59 @@ class LanguageServerConfigPage(GeneralConfigPage):
...
+        self.completion_hint_box = newcb(
+            _("Show completion details"),
+            'completions_hint',
+            section='editor')
+        self.automatic_completion_box = newcb(
+            _("Show completions on the fly"),
+            'automatic_completions',
+            section='editor')
+        self.completions_after_characters = self.create_spinbox(
+            _("Show automatic completions after characters:"), None,
```
Comment (ccordoba12) [inline spyder/plugins/completion/languageserver/confpage.py]: Let's follow RStudio here and change this text to be

    Show automatic completions after characters entered:

File: spyder/plugins/completion/languageserver/confpage.py [hunk 4]
```diff
@@ -1053,20 +1099,34 @@ def setup_page(self):
 
         # --- Tabs organization ---
         self.tabs = QTabWidget()
-        self.tabs.addTab(self.create_tab(basic_features_group, advanced_group),
-                    _('Introspection'))
+        self.tabs.addTab(self.create_tab(completion_widget),
+                         _('Completion'))
+        self.tabs.addTab(self.create_tab(introspection_group, advanced_group),
+                         _('Introspection'))
```
Comment (ccordoba12) [inline spyder/plugins/completion/languageserver/confpage.py]: Please move this tab to be after the `Linting` one (as I mentioned before) because it's very strange that the entry in Preferences is called "Completion and linting" but not seeing those two tabs one after the other here.

Comment (CAM-Gerlach) [reply spyder/plugins/completion/languageserver/confpage.py]: The rationale here is likely that which I outlined on #10134 : given code analysis ("linting"), code style analysis and docstring style analysis are closely connected at least from a user's perspective, it makes sense to keep them together.

If you insist on linting second after completion (which from the perspective of a user I really don't see as mattering all that much, at least compared to grouping the three real-time code analysis features together), Introspection could be moved after docstring style, or you could consider renaming the preferences section to a broader term more inclusive of its contents (though I'm not sure what, exactly).

**Comment by goanpeca** [comment]
@CAM-Gerlach thanks for testing this out!

>There's one thing, though—since you already have the option implemented, can you allow a minimum delay of 0 ms rather than 200 ms, for users who prefer maximum responsiveness in their OTF autocompletion? As far as I'm aware there's no technical reason to restrict it (given it was 0 before, and its already configurable so it doesn't increase complexity any further) and some users may not want any delay or very little (I had it set to 100 ms in Rstudio) though 300 ms is more than fine as the default.

Sure sounds fine.

**Comment by CAM-Gerlach** [comment]
Thanks @goanpeca !

File: spyder/plugins/completion/languageserver/confpage.py (hunk 4 above)
Comment (goanpeca) [reply spyder/plugins/completion/languageserver/confpage.py]: I think both are valid points!

**Comment by goanpeca** [comment]
>This looks pretty good, thanks @goanpeca! I left you some minor comments and I guess you also need to do a rebase on top of master because I already merged PR #10253.

Fixed @ccordoba12!
